keps/sig-cluster-lifecycle/kubeadm/20190424-kubeadm-for-windows.md

```diff
@@ -0,0 +1,293 @@
+
+This proposal plans for FlannelD as the default option. Currently, FlannelD has to be started before the kube-proxy Windows service is started. FlannelD creates an HNS network on the Windows host, and kube-proxy will crash if it cannot find the network. This should be fixed in the scope of this project so that kube-proxy will wait until the network comes up. Therefore, kube proxy can be started at any time. 
+
+However, if FlannelD is deployed in VXLAN (Overlay) mode, then we need to rewrite the KubeProxyConfiguration with the correct Overlay specific values, and kube-proxy will need to read this config again.
+
+If the users wish to use a different plugin they will technically opt-out of the supported setup for kubeadm based Windows worker nodes in the Alpha release.
+
+Long term, any CNI plugin should be supported for kubeadm based Windows worker nodes.
+
+### Risks and Mitigations
+
+**Risk**: Versioning of the wrapper script can become complicated
+
+Versioning of the script per-Kubernetes version can become a problem if a certain new version diverges in terms of download URLs, flags and configuration.
+
+*Mitigation*: Use git branches to version the script in the repository where it is hosted.
+
+**Risk**: The wrapper script is planned to act as both a downloader and runner of the downloader binaries, which might cause scope and maintenance issues.
+
+*Mitigation*: Use separate scripts, the first one downloads the binaries and the wrapper/runner script then setups the environment. The user then executes the wrapper script.
+
+The initial plan is to give the single script method a shot with different arguments that will execute the different stages (downloading, setting up the environment, deploying the CNI).
+
+**Risk**: Flannel or kube-proxy require special configuration that kubeadm does not handle.
+
+*Mitigation*: Allow the user to pass custom configuration files that the wrapper script can feed into the components in question.
+
+**Risk**: Failing or missing preflight checks on Windows
+
+*Mitigation*: The existing kubeadm codebase already has good abstraction in this regard. Still, a PR that makes some non-intrusive adjustments in _windows.go files might be required.
+
+**Risk**: Permissions on Windows paths that kubeadm generates can pose a security hole.
+
+kubeadm creates directories using MakeAll() and such directories are strictly Linux originated for the time being - such as /etc.
+
+On Windows, the creation of such a path can result in sensitive files to be exposed without the right permissions.
+
+*Mitigation*: Manage the preemptive creation of sensitive paths from the wrapper script. Make such paths require Administrator level access.
```

Comment:
I am not convinced that we should outsource that to the script. Providing a `SecureMakeAll()` func in kubeadm, that ensures secure enough permissions on both Windows & Linux, and replacing usage of `MakeAll()` is one simple solution. I am sure, that there are others as well.

Reply:
the abstraction of secure paths would require kubeadm in the case of Windows to import some golang/x or make calls to the tool that changes permissions on Windows.
i don't like having this in kubeadm for the Alpha.

Reply:
What is the resolution on this?

Reply:
"from the wrapper script" is the initial idea, but subject of change.